crates/ruff_db/src/diagnostic/render/json.rs

```diff
@@ -0,0 +1,194 @@
+use serde::{Serialize, Serializer, ser::SerializeSeq};
+use serde_json::{Value, json};
+
+use ruff_diagnostics::Edit;
+use ruff_notebook::NotebookIndex;
+use ruff_source_file::{LineColumn, OneIndexed, SourceCode};
+use ruff_text_size::Ranged;
+
+use crate::diagnostic::Diagnostic;
+
+use super::FileResolver;
+
+pub(super) fn diagnostics_to_json_value<'a>(
+    diagnostics: impl IntoIterator<Item = &'a Diagnostic>,
+    resolver: &dyn FileResolver,
+) -> Value {
+    let messages: Vec<_> = diagnostics
+        .into_iter()
+        .map(|diag| diagnostic_to_json_value(diag, resolver))
+        .collect();
+    json!(messages)
+}
+
+pub(super) fn diagnostic_to_json_value(message: &Diagnostic, resolver: &dyn FileResolver) -> Value {
+    let span = message.primary_span_ref();
+    let filename = span.map(|span| span.file().path(resolver));
+    let range = span.and_then(|span| span.range());
+    let diagnostic_source = span.map(|span| span.file().diagnostic_source(resolver));
+    let source_code = diagnostic_source
+        .as_ref()
+        .map(|diagnostic_source| diagnostic_source.as_source_code());
+    let notebook_index = span.and_then(|span| resolver.notebook_index(span.file()));
+
+    let fix = message.fix().map(|fix| {
+        json!({
+            "applicability": fix.applicability(),
+            "message": message.suggestion(),
+            "edits": &ExpandedEdits {
+                edits: fix.edits(),
+                source_code: source_code.as_ref(),
+                notebook_index: notebook_index.as_ref()
+            },
+        })
+    });
+
+    let mut start_location = None;
+    let mut end_location = None;
+    let mut noqa_location = None;
+    let mut notebook_cell_index = None;
+    if let Some(source_code) = source_code {
+        noqa_location = message
+            .noqa_offset()
+            .map(|offset| source_code.line_column(offset));
+        if let Some(range) = range {
+            let mut start = source_code.line_column(range.start());
+            let mut end = source_code.line_column(range.end());
+            if let Some(notebook_index) = notebook_index {
+                notebook_cell_index =
+                    Some(notebook_index.cell(start.line).unwrap_or(OneIndexed::MIN));
+                start = notebook_index.translate_line_column(&start);
+                end = notebook_index.translate_line_column(&end);
+                noqa_location =
+                    noqa_location.map(|location| notebook_index.translate_line_column(&location));
+            }
+            start_location = Some(start);
+            end_location = Some(end);
+        }
+    }
+
+    json!({
+        "code": message.secondary_code(),
+        "url": message.to_ruff_url(),
+        "message": message.body(),
+        "fix": fix,
+        "cell": notebook_cell_index,
+        "location": start_location.map(location_to_json),
+        "end_location": end_location.map(location_to_json),
```

Comment:
Making `location` and `end_location` `Option` is a breaking change. It's a bit hard to tell if you made other breaking changes to the schema. Please double check that you didn't.

I think the change I would make here is to default to `TextRange::default` if the range is missing and preview mode is disabled (we can use an `Option` in preview mode).